File: creational/typescript/singleton_structured/configuration_manager.ts
export class ConfigurationManager {
  private static instance: ConfigurationManager | null = null;
  private config: Map<string, string> = new Map();
  
  private constructor(initialConfig?: Record<string, string>) {
    console.log('ConfigurationManager: Initializing...');
    if (initialConfig) {
      Object.entries(initialConfig).forEach(([key, value]) => {
        this.config.set(key, value);
      });
    }
  }
  
  public static getInstance(initialConfig?: Record<string, string>): ConfigurationManager {
    if (!ConfigurationManager.instance) {
      ConfigurationManager.instance = new ConfigurationManager(initialConfig);
    }
    return ConfigurationManager.instance;
  }
  
  public set(key: string, value: string): void { this.config.set(key, value); }
  public get(key: string): string | undefined { return this.config.get(key); }
  
  public displayConfig(): void {
    console.log('Current Configuration:');
    this.config.forEach((value, key) => {
      console.log(`  ${key}: ${value}`);
    });
  }
}

File: creational/dart/singleton_structured/configuration_manager.dart
class ConfigurationManager {
  static ConfigurationManager? _instance;
  
  final Map<String, String> _config = {};
  
  ConfigurationManager._({Map<String, String>? initialConfig}) {
    print('ConfigurationManager: Initializing...');
    if (initialConfig != null) {
      _config.addAll(initialConfig);
    }
  }
  
  static ConfigurationManager getInstance({Map<String, String>? initialConfig}) {
    _instance ??= ConfigurationManager._(initialConfig: initialConfig);
    return _instance!;
  }
  
  void set(String key, String value) {
    _config[key] = value;
  }
  
  String? get(String key) {
    return _config[key];
  }
  
  void displayConfig() {
    print('Current Configuration:');
    _config.forEach((key, value) {
      print('  $key: $value');
    });
  }
}
